### ISSUE TITLE
Bug: prefer-const flagged for incorrectly

### ISSUE THREAD

**Issue by JoshMcCullough** [eslint/eslint#17192]
### Environment

Node version: 18.4.0
npm version: 8.12.1
Local ESLint version: 8.40.0
Global ESLint version: N/A
Operating System: Manjaro Linux 22.1.1

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
module.exports = {
  extends: [
    'eslint:recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:@typescript-eslint/recommended-requiring-type-checking',
  ],
  parser: '@typescript-eslint/parser',
  plugins: [
    '@typescript-eslint',
  ],
  root: true,
  overrides: [
    {
      files: ['*.ts'],
      parserOptions: {
        project: 'tsconfig.json',
      },
      rules: {
        '@typescript-eslint/prefer-nullish-coalescing': 'error',
      },
    },
  ],
};

```
</details>

```js
async function init(testSuiteName: string, configureRepo: ((repo: RepositoryWithFiles) => void) | null): Promise<Context> {
  const ghConfig: TestGithubConfig = {
    // set up default config
  };
  const repo = ghConfig.repo.test;
  let gh: MockGithub;

  // allow for conf to be tweaked
  configureRepo?.(repo);

  // assume this constructor expects the config to be ready-to-go
  gh = new MockGithub(ghConfig, workDir);
  gh.setup();
  
  // additional code

  return gh;
}
```

### What did you expect to happen?

Note that `gh` has to be constructed after  `configureRepo` is called, so the config is complete before `gh` is instantiated. The obvious alternative is to move the declaration of `gh`  to after `configureRepo`:

```js
  // all the other code here

  configureRepo?.(repo);

  const gh = new MockGithub(ghConfig, workDir);

  gh.setup();

  return gh;
```

But this violates the "variable at top of the block" principle -- e.g. not declaring variables mid-block.

And unfortunately it's not valid to declare a `const` variable w/out initializing it at the same time. (E.g. like you can with Java.)

### What actually happened?

ESlint gives the error:

```
error    'gh' is never reassigned. Use 'const' instead    prefer-const
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiY2xhc3MgVGhpbmcge1xuICBibGFoID0gZmFsc2U7XG4gIFxuICBjb25zdHJ1Y3Rvcihjb25maWcpIHtcbiAgICB0aGlzLmJsYWggPSBjb25maWcuYmxhaDtcbiAgICB0aGlzLmRvU29tZXRoaW5nKGNvbmZpZylcbiAgfVxuXG4gIGRvU29tZXRoaW5nKGNvbmZpZykge1xuICAgIGNvbmZpZy5maWxlcy5mb3JFYWNoKGYgPT4ge1xuICAgICAgLy8gZS5nLiBtYWtlIHN1cmUgdGhlIGZpbGUgZXhpc3RzXG4gICAgICBjb25zb2xlLmxvZyhmKTtcbiAgICB9KVxuICB9XG59XG5cbmZ1bmN0aW9uIGluaXRUaGluZyhjb25maWd1cmUpIHtcbiAgY29uc3QgY29uZmlnID0ge1xuICAgIGJsYWg6IHRydWUsXG4gICAgZmlsZXM6IFt7XG4gICAgICBuYW1lOiAnc29tZS1maWxlLnR4dCcsXG4gICAgICBwYXRoOiAnL3BhdGgvdG8vc29tZS1maWxlLnR4dCcsXG4gICAgfV0sXG4gIH07XG4gIGxldCB0aGluZztcblxuICBjb25maWd1cmUoY29uZmlnKTtcblxuICB0aGluZyA9IG5ldyBUaGluZyhjb25maWcpO1xuICB0aGluZy5zZXR1cCgpO1xuXG4gIHJldHVybiB0aGluZztcbn1cblxuLy8gbWFrZSBhIHRoaW5nXG5jb25zdCB0ID0gaW5pdFRoaW5nKGNvbmZpZyA9PiB7XG4gIGNvbmZpZy5maWxlcy5wdXNoKHtcbiAgICBuYW1lOiAnc29tZS1vdGhlci1maWxlLnR4dCcsXG4gICAgcGF0aDogJy9wYXRoL3RvL3NvbWUtb3RoZXItZmlsZS50eHQnLFxuICB9KVxufSkiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InByZWZlci1jb25zdCI6WyJlcnJvciJdfSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

Arguments against my proposal this might be:
1. _The `init` function should be broken into an additional function to handle the creation of the `gh` instance,_ -- Possibly, but this is just one example, and it is not always correct so split a function just to get around a lint rule.
2. _You should just declare the `const gh` mid-block where it's initialized._ -- I disagree, and will admit that the need to assign a const later in a block points to #1 (splitting the function). E.g. if you find that you need to declare a variable mid-block then you should probably break the function up.
3. _A constructor shouldn't be affect-less, so you should fix the constructor of the `gh` instance so it doesn't do anything with the config._ -- Perhaps, but not if it's a 3rd party library.

**Comment by B2o5T** [comment]
I don’t see issues with the rule, your proposal will lead to future issues with false negatives cases for this rule, so this battle will never end

Just go with step 2

or just use eslint-disable comment for your code style

**Comment by Rec0iL99** [comment]
cc @eslint/eslint-tsc

**Comment by nzakas** [comment]
The rule is behaving as designed. If you want to allow the code to be written in the way you describe, you'll need use some disable comments.
